packages/riipen-ui/src/components/TableGenerator.jsx

```diff
@@ -0,0 +1,277 @@
+  }
+
+  getLinkedStyles = () => {
+    const theme = this.context;
+
+    return css.resolve`
+      .container {
+        border: 1px solid ${theme.palette.grey[400]};
+        border-radius: 4px;
+        display: inline-block;
+        width: 100%;
+      }
+
+      .centeredTable {
+        margin: auto;
+        min-width: 50%;
+        width: auto;
+      }
+
+      .noPadding {
+        padding: 0;
+      }
+
+      td,
+      th {
+        max-width: 400px;
+        vertical-align: middle;
+      }
+    `;
+  };
+
+  getCellAlignment(cell) {
+    return cell?.align || "left";
+  }
+
+  static contextType = ThemeContext;
+
+  updateWindowDimensions = () => {
+    const theme = this.context;
+    const windowWidth = window.innerWidth;
+    const breakpoint = theme.breakpoints[this.props.mobileBreakpoint];
+    this.setState({
+      isMobile: windowWidth <= breakpoint
+    });
+  };
+
+  renderTableHeader() {
+    const { isMobile } = this.state;
+    const { columns } = this.props;
+
+    const linkedStyles = this.getLinkedStyles();
+
+    return (
+      <TableHeader>
+        <TableRow border>
+          {columns.map((col, i) => (
+            <TableHeaderCell
+              classes={[linkedStyles.className]}
+              key={`header-column-${i}`}
+              align={this.getCellAlignment(col)}
+            >
+              {col?.header(isMobile)}
+            </TableHeaderCell>
+          ))}
+        </TableRow>
+      </TableHeader>
+    );
+  }
+
+  renderDefault() {
+    const { hover, data, columns } = this.props;
+    const { isMobile } = this.state;
+
+    const linkedStyles = this.getLinkedStyles();
+
+    return (
+      <React.Fragment>
+        {this.renderTableHeader()}
+        <TableBody>
+          {data.map((row, i) => {
+            return (
+              <TableRow hover={hover} key={`default-row-${i}`}>
+                {columns.map((col, j) => (
+                  <TableCell
+                    classes={[linkedStyles.className]}
+                    key={`default-cell-${i}-${j}`}
+                    align={this.getCellAlignment(col)}
+                  >
+                    {col?.cell(row, isMobile)}
+                  </TableCell>
+                ))}
+              </TableRow>
+            );
+          })}
+        </TableBody>
+      </React.Fragment>
+    );
+  }
+
+  renderMobile() {
+    const { hover, columns, data } = this.props;
+    const { isMobile } = this.state;
+
+    const linkedStyles = this.getLinkedStyles();
+
+    const bodyColumns = Object.assign([], columns);
+    let mobileHeader;
+    let mobileFooter;
+
+    const mobileHeaderIndex = bodyColumns.findIndex(x => x.mobileHeader);
+    if (mobileHeaderIndex !== -1) {
+      mobileHeader = bodyColumns.splice(mobileHeaderIndex, 1)[0];
+    }
+
+    const mobileFooterIndex = bodyColumns.findIndex(x => x.mobileFooter);
+    if (mobileFooterIndex !== -1) {
+      mobileFooter = bodyColumns.splice(mobileFooterIndex, 1)[0];
+    }
```

Comment:
personal preference: I don't like splice. it mutates and hardly ever gets used (I had to look at the mdn docs to see what the second param was, and what it returns!). In general I find it makes code harder to follow then it otherwise would be.

If I got the logic right you can do something like:

```javascript
const mobileHeader = columns.find(x => x.mobileHeader);
const mobileFooter = columns.find(x => x.mobileFooter);
const bodyColumns = columns.filter(x => !x.mobileFooter && !x.mobileHeader);
```

Which IMO is a bit more clear.

Reply:
you see, I was thinking the same thing... and for some reason [].filter just slipped my mind.. It was late lol